src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php

```diff
@@ -54,9 +55,15 @@ protected function checkAuthentication(UserInterface $user, UsernamePasswordToke
                 throw new BadCredentialsException('The presented password cannot be empty.');
             }
 
-            if (!$this->encoderFactory->getEncoder($user)->isPasswordValid($user->getPassword(), $presentedPassword, $user->getSalt())) {
+            $encoder = $this->encoderFactory->getEncoder($user);
+
+            if (!$encoder->isPasswordValid($user->getPassword(), $presentedPassword, $user->getSalt())) {
                 throw new BadCredentialsException('The presented password is invalid.');
             }
+
+            if ($this->userProvider instanceof PasswordUpgraderInterface && method_exists($encoder, 'needsRehash') && $encoder->needsRehash($user->getPassword())) {
+                $this->userProvider->upgradePassword($user, $encoder->encodePassword($presentedPassword, $user->getSalt()));
```

Comment:
This gives this method two responsibilies: checking the password and upgrading it if needed. Maybe this should be done at some higher -orchistration- level?

Reply:
You mean the checkAuthentication method? I've no better idea, can you elaborate?

Reply:
Yes, the `checkAuthentication` method.
So instead of doing this inside the `checkAuthentication` do this in the code that calls `checkAuthentication`, so move it one level up to decouple the behaviours:

```
if ($provider->checkAuthentication(...)) {
    $provider->upgradePassword(...);
}
```

Reply:
I'll have a look but I think what misses in your example is the encoder: the "if" must involve it before calling upgrade. Yet the encoder is internal detail of the provider.

Reply:
> encoder is internal detail of the provider

I confirm - and so is the cleartext password btw. Same for guard btw.
Dead end to me.

Reply:
Seems to me it could be moved to `\Symfony\Component\Security\Core\Authentication\Provider\UserAuthenticationProvider`. That way it's part of the template method so all AuthenticationProviders can benefit from it, instead of being part of a particular implementation.

```
public function authenticate(TokenInterface $token)
{
    // ...
    try {
        $this->userChecker->checkPreAuth($user);
        $this->checkAuthentication($user, $token);
        if ($this instanceof PasswordUpgradingAuthenticationProvider) {
            // optional: if ($this->needsPasswordRehash($user, $token)) {
            $this->upgradeUserPassword($user, $token);
            // optional: }
        }
        $this->userChecker->checkPostAuth($user);
    } catch (BadCredentialsException $e) {
        if ($this->hideUserNotFoundExceptions) {
            throw new BadCredentialsException('Bad credentials.', 0, $e);
        }
        throw $e;
    }
```

and then `\Symfony\Component\Security\Core\Authentication\Provider\DaoAuthenticationProvider` would implement `PasswordUpgradingAuthenticationProvider`:

```
class DaoAuthenticationProvider extends UserAuthenticationProvider implements PasswordUpgradingAuthenticationProvider
{
   // ...
   public function upgradeUserPassword(UserInterface $user, UsernamePasswordToken $token): void
   {
      if ($this->userProvider instanceof PasswordUpgraderInterface && method_exists($encoder, 'needsRehash') && $encoder->needsRehash($user->getPassword())) {
          $this->userProvider->upgradePassword($user, $encoder->encodePassword($token->getCredentials(), $user->getSalt()));
      }
   }
}
```

Reply:
> optional: if ($this->needsPasswordRehash($user, $token)) {

yes but no :) this is not optional: this is part of the critical lifecycle of password migrations.
Not all auth providers have a concept of encoders. Only those should know and care about migrations. Said another way, your initial issue was SRP for the method, this proposal is moving the issue somewhere else. Let's say there is no SRP issue in the first place and everything is fine: checkAuthentication is the correct place to wire opportunistic password upgrades to me.

Reply:
My issue was with CQS, not SRP. My issue with SRP was in the naming of the `ChainPasswordEncoder`.

My issue here is still CQS, and I think it would be solved by my proposal.

Reply:
Sorry it wouldn't, see my objections about encoders. See also the guard interface.
Opportunistic password upgrades don't align to CQS here to me.

Reply:
But please submit a PR on my fork if I'm missing the point :)